Google.Api.Gax.Grpc/Rest/HttpRulePathPattern.cs

```diff
@@ -9,158 +9,262 @@
+                if (braceEnd != -1)
                 {
-                    string literal = pattern.Substring(lastEnd, start - lastEnd);
-                    segments.Add(HttpRulePathPatternSegment.CreateFromLiteral(literal));
+                    throw new ArgumentException($"Resource pattern '{pattern}' is invalid: unmatched }}");
                 }
-
-                // `startIndex: 1` for `{`; `match.Value.Length - 2` for `{}`
-                string boundFieldPath = match.Value.Substring(1, match.Value.Length - 2);
-                string boundFieldPattern = "";
-
-                int endOfName = match.Value.IndexOf('=');
-                if (endOfName != -1)
-                {
-                    // `startIndex: 1` for `{`; `endOfName - 1` for `=`
-                    boundFieldPath = match.Value.Substring(1, endOfName - 1);
-                    // `endOfName + 1` for `=`; `match.Value.Length - 3` for `={}`
-                    boundFieldPattern = match.Value.Substring(endOfName + 1, match.Value.Length - 3 - boundFieldPath.Length);
-                }
-
-                segments.Add(HttpRulePathPatternSegment.CreateFromBoundField(boundFieldPath, boundFieldPattern, requestDescriptor));
-                lastEnd = match.Index + match.Length;
+                break;
             }
-            if (lastEnd != pattern.Length)
+            // Either the closing brace comes before the opening one, or doesn't exist.
+            if (braceEnd < braceStart)
             {
-                string literal = pattern.Substring(lastEnd);
-                segments.Add(HttpRulePathPatternSegment.CreateFromLiteral(literal));
+                throw new ArgumentException($"Resource pattern '{pattern}' is invalid: unmatched {{");
             }
-            return new HttpRulePathPattern(segments);
+            if (braceStart != currentIndex)
+            {
+                string literal = pattern.Substring(currentIndex, braceStart - currentIndex);
+                segments.Add(new LiteralSegment(literal));
+            }
+            segments.Add(new FieldSegment(pattern.Substring(braceStart, braceEnd - braceStart + 1), requestDescriptor));
+            currentIndex = braceEnd + 1;
+        }
+        if (currentIndex != pattern.Length)
+        {
+            segments.Add(new LiteralSegment(pattern.Substring(currentIndex)));
         }
+        return new HttpRulePathPattern(segments);
+    }
 
-        internal string Format(IMessage message) => string.Join("", _segments.Select(segment => segment.Format(message)));
+    /// <summary>
+    /// Attempts to format the path with respect to the given request,
+    /// returning the formatted segment or null if formatting failed due to the request data.
+    /// </summary>
+    internal string TryFormat(IMessage message)
+    {
+        StringBuilder builder = new StringBuilder();
+        var segments = string.Join("", _segments.Select(segment => segment.TryFormat(message)));
```

Comment:
I don't think you are using this anymore.

Reply:
Thanks - will fix in an extra tidy-up commit.